cockpit:machines: Display VM graphics console

mareklibra opened the pull request:
With this patch, graphics console(s) can be retrieved for a VM.

Console connection details are retrieved from VM's domail.xml.
To play around, please be sure the test VM is created with appropriate external IP set for the console.
Example for `virt-install`:
  `virt-install --graphics spice,listen=[external host IP]`

So far tested for:
- VNC (virt-viewer)
- SPICE (virt-viewer)
- in browser console via noVnc

The virt-viewer is invoked by downloading a .vv file of `application/x-virt-viewer` mime type.

Following other projects were taken into consideration for the console-in-browser feature:
- spice-html5
- [broadwayd](https://dummdida.tumblr.com/post/131044606675/view-your-gtk3-app-or-vm-on-the-web)

The `broadwayd` is cool but heavy-weight, requiring a lot of client-side dependencies on server.

The `spice-html5` is working for basic scenarios but has multiple limitations so we can't consider it as a full-feature solution. In addition, community support is very limited.

-----
The cockpit-machines newly sets
    "content-security-policy": "default-src 'self' '**unsafe-inline**' data:"
to make the no-vnc and download of dynamically generated files (the .vv) working.

 - [x] fix or disable inline VNC with plain http://
 - [x] integration tests
 - [ ] get design sign-off from @garrett 
 - [x] make VNC window larger and/or resizable to avoid scrolling inside the widget
 - [x] address `check-machines TestMachines.testBasic failure on ubuntu-*` test failure

mareklibra commented:
Please see the screenshot.

The graphics console can be retrieved via button rendered along the other "active" actions (like "Shut Down").
If a VM has defined multiple consoles, they are listed in a drop-down button list (using button with a caret, same approach as for the "Shut Down [Force]").

Design questions:
- Should we name the consoles according to it's type? (like recently implemented "spice" or "vnc"). Any other idea? The user needs to differentiate between them.
- Are html buttons nice-enough for the growing list of active actions? What about using icons? Potential change would be implemented in a follow-up PR.

mareklibra commented:
![consoles](https://cloud.githubusercontent.com/assets/17194943/23250832/0a5a19cc-f9ab-11e6-97bb-9b2e3bf50f8d.png)

stefwalter commented:
>  The spice-html5 is working for basic scenarios but has multiple limitations so we can't consider it as a full-feature solution. In addition, community support is very limited.

The Fleet Commander guys seem to have SPICE HTML working really well. I would suggest pinging @olivergs to figure out where they're using it from.

michalskrivanek commented:
@stefwalter we have spice-html5 integrated within oVirt platform for cca 3 years but do not really have a good experience. It is too buggy and lacks features of virt-viewer. We kept it only as a workaround, or a complementary method, for people who can't use or install virt-viewer, but even there the typical "solution" was to reconfig VMs to VNC and use novnc.

I think it's a good idea to support both browser and regular clients, both have their advantages.

stefwalter commented:
> I think it's a good idea to support both browser and regular clients, both have their advantages.

Yes, all these options could be in the drop down. But we should have a way to do it from HTML itself (without all the broadway stuff on the server) whether that's SPICE-HTML5 or noVNC. I don't mind if broadway or virt-viewer are also present.

michalskrivanek commented:
@stefwalter, about the drop down - not sure if you're familiar with oVirt's client side console options dialog (http://deepawake.com/ovirt-engine/docs/manual/en_US/html/Virtual_Machine_Management_Guide/images/5679.png), but we would eventually love to see such functionality. There are quite a few client side options you want to set, per user, per VM, including the selection of best client to invoke.  

As for HTML native client in cockpit, If I may I would recommend to start with VNC only via novnc. It is stable, small foot print, reasonably feature complete. There are some issues with keyboard mappings, but nothing major.

olivergs commented:
Hey!

Right now we discarded VNC as client and started to use SPICE due to custom
channels feature. Some months ago I sent a patch for including custom
channels support to spice-html5. You can test it and check it following the
simple instructions in my personal blog.

Lately spice-html5 has received some updates but I haven't updated the
client for fleet commander yet so I can not tell about performance
enhancements.

If you want to check how it is working in Fleet Commander you can check out
source code. More specifically this file:
https://github.com/fleet-commander/fc-admin/blob/master/admin/cockpit/fleet-commander-admin/js/fcspiceclient.js

We are using upstream version without modifications.

Hope this help.

Cheers!



On Thu, Feb 23, 2017 at 3:22 PM, michalskrivanek <notifications@github.com>
wrote:

> @stefwalter <https://github.com/stefwalter>, about the drop down - not
> sure if you're familiar with oVirt's client side console options dialog (
> http://deepawake.com/ovirt-engine/docs/manual/en_US/html/
> Virtual_Machine_Management_Guide/images/5679.png), but we would
> eventually love to see such functionality. There are quite a few client
> side options you want to set, per user, per VM, including the selection of
> best client to invoke.
>
> As for HTML native client in cockpit, If I may I would recommend to start
> with VNC only via novnc. It is stable, small foot print, reasonably feature
> complete. There are some issues with keyboard mappings, but nothing major.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockpit-project/cockpit/pull/5932#issuecomment-282021803>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABX4ofiKn2PdsC4UAQ1uZJZeeC1Dhcs_ks5rfaQzgaJpZM4MJrie>
> .
>



-- 
Oliver Gutierrez
Associate Software Engineer - Desktop Management tools
Red Hat

olivergs commented:
Forgot to link the blog: https://matrixhas.me/en/spice-port-feature/

On Thu, Feb 23, 2017 at 5:15 PM, Oliver Gutierrez <ogutierrez@redhat.com>
wrote:

> Hey!
>
> Right now we discarded VNC as client and started to use SPICE due to
> custom channels feature. Some months ago I sent a patch for including
> custom channels support to spice-html5. You can test it and check it
> following the simple instructions in my personal blog.
>
> Lately spice-html5 has received some updates but I haven't updated the
> client for fleet commander yet so I can not tell about performance
> enhancements.
>
> If you want to check how it is working in Fleet Commander you can check
> out source code. More specifically this file: https://github.com/fleet-
> commander/fc-admin/blob/master/admin/cockpit/fleet-commander-admin/js/
> fcspiceclient.js
>
> We are using upstream version without modifications.
>
> Hope this help.
>
> Cheers!
>
>
>
> On Thu, Feb 23, 2017 at 3:22 PM, michalskrivanek <notifications@github.com
> > wrote:
>
>> @stefwalter <https://github.com/stefwalter>, about the drop down - not
>> sure if you're familiar with oVirt's client side console options dialog (
>> http://deepawake.com/ovirt-engine/docs/manual/en_US/html/Vi
>> rtual_Machine_Management_Guide/images/5679.png), but we would eventually
>> love to see such functionality. There are quite a few client side options
>> you want to set, per user, per VM, including the selection of best client
>> to invoke.
>>
>> As for HTML native client in cockpit, If I may I would recommend to start
>> with VNC only via novnc. It is stable, small foot print, reasonably feature
>> complete. There are some issues with keyboard mappings, but nothing major.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/cockpit-project/cockpit/pull/5932#issuecomment-282021803>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABX4ofiKn2PdsC4UAQ1uZJZeeC1Dhcs_ks5rfaQzgaJpZM4MJrie>
>> .
>>
>
>
>
> --
> Oliver Gutierrez
> Associate Software Engineer - Desktop Management tools
> Red Hat
>



-- 
Oliver Gutierrez
Associate Software Engineer - Desktop Management tools
Red Hat

dperpeet commented:
this requires a rebase

mareklibra commented:
Rebased

mareklibra commented:
The noVNC is added.

Pushed to get feedback, but still work in progress. 

Aspects to highlight:
- rendered to an iframe
- noVNC modules are not bundled - loaded within multiple requests (Why? See vnc.html:64)
- manifest.json: "content-security-policy": "default-src 'self' 'unsafe-inline' 'unsafe-eval' data:"

TODO:
 - user can resize iframe (drag-n-drop)
 - considering moving the status bar outside of the console area

mareklibra commented:
![listavailableconsoles](https://cloud.githubusercontent.com/assets/17194943/24092287/b2b6f3f0-0d4d-11e7-835c-0a714b0f6f4c.png)

mareklibra commented:
![novnc](https://cloud.githubusercontent.com/assets/17194943/24092293/baf1d67a-0d4d-11e7-8591-88f0c92e7a24.png)

mareklibra commented:
@andreasn , @garrett , may I ask you for feedback, please?

mareklibra commented:
Pull request description is updated to reflect progress with the noVnc

mareklibra commented:
Rebased

garrett commented:
Hi @mareklibra! I kind of have this working from your branch locally — enough to see the VMs in a list and inspect them, complete with a console tab — but I only see a link called "SPICE in Desktop Viewer" that doesn't work. It's probably some misconfiguration on my end (perhaps even related to setting up my replacement work laptop from scratch). As such, I'll base the suggestions below on what I can see locally and what is posted above in this PR.

---

From what I understand of the feature, based on the screenshots above, I think an optimal way of displaying the UI this is to use an in-browser solution by default. It would be powered by SPICE in-browser, but fall back to VNC also in-browser if SPICE cannot be used for whatever reason. 

I'd then suggest combining the close desktop SPICE or VNC session and the view in desktop SPICE/VNC viewer to "Switch to Desktop Viewer", which then would close the in-page session and show instructions on how to view the session remotely from the desktop, possibly with side-by side instructions for both SPICE and VNC viewers.

When in the desktop viewing mode, there would be a way to switch back, probably called something like "Switch to Web-Based Viewer" in the place of the "Switch to Desktop Viewer".

As most people would switch either because they prefer a desktop viewing experience or because of technical issues with interacting with the VMs in-browser, I think it would be a good idea to remember the toggle state and show new console tabs in previously chosen state (if not across sessions, then at least in the same session).

---

## Summary

- Default to in-browser viewing
- Allow someone to easily switch between in-browser and desktop clients with a toggle link
- Desktop clients mode would have instructions (possibly including a link) on how to view the machine in SPICE and/or VNC (whatever applies)
- Remember the state of web/desktop viewing and apply that to newly opened console tabs (at least within the Cockpit session)

jelkosz commented:
Hi all,

sorry for joining the party late. Im a developer on the oVirt.

Please note that there are also other settings which can be configured for SPICE and VNC and it might be useful to configure them also here:
- SPICE:
  - Map ctrl+alt+del shortcut to ctrl+alt+end (quite useful for windows machines)
  - Enable USB Auto-Share (e.g. pass through the USB device from the client machine inside to the virtual machine)
  - Open in Full Screen
  - Enable SPICE Proxy (in oVirt this is only a checkbox but it expects the SPICE proxy is already configured for this VM. So maybe not so useful for cockpit)

- VNC
  - Map ctrl+alt+del shortcut to ctrl+alt+end

@garrett do you think this options make sense for the cockpit use case as well?

michalskrivanek commented:
I'd just add for more context:
- there are more generally useful client-side options (WAN
optimization, preferred encoding/compression, etc)
- they are mostly relevant/supported in virt-viewer, very few are
supported for browser-based clients.

mareklibra commented:
@garret, thanks for your comment, I updated the code accordingly with the exception, there's recently no in-browser Spice support (like spice-html5). But the code is open for its addition later. Same as the RDP is missing for now.
Please have a look at recent screenshots.

Behavior if VNC is defined:
 - the noVNC frame is rendered by default when Console subtab is clicked.
 - user can change to "Desktop Spice" or "Desktop VNC" and back
 - if any desktop console is selected, descriptive text is rendered with steps to resolve potential issues

Behavior if no console with in browser support is defined (recently means SPICE without VNC):
 - the "descriptive" text is rendered as above
 - user can click "Desktop SPICE", download is not fired automatically

Troubleshooting description for the "desktop variant" can be customized by the external provider, since i.e. oVirt/RHV have specifics.
![01_defaultinbrowser](https://cloud.githubusercontent.com/assets/17194943/24543304/6d9ad060-15ff-11e7-8c9e-e27bbace3272.png)
![02_desktopconsoleselected](https://cloud.githubusercontent.com/assets/17194943/24543313/74ec233c-15ff-11e7-883e-0cc84f77f8fe.png)

mareklibra commented:
Let's not skip comments/question of @jelkosz and @michalskrivanek above.

mareklibra commented:
@dperpeet, @stefw, any idea how to resolve the issue with 'unpatchable machines/include/keysymdef.js file'? See failing semaphoreci, please.


I tried to bundle the no-vnc distributables to a single .js, but failed - reasons are listed in the vnc.html:62 within this PR.

So, various no-vnc .js files are dynamically loaded from "machines/include/*" URL.
To do so, I added symlink "pkg/machines/include" pointed to bower_componentns/no-vnc/include, where the keysymdef.js resides.

If we stay with this "non-bundled" approach, we might within pkg/Makefile.am:
- avoid the symlink by explicitly copying the no-vnc files
- process selected files (means break lines in the keysymdef.js)

Any better idea or even example, please?

garrett commented:
@mareklibra: Nice. It's a step in the right direction.

What does clicking on "console client resources" do?

I'd also like to have the no-VNC-available screen to be more like a normal state (as normal as having VNC working) versus being more of an error/fallback state. To do this I'd suggest:

1. Including the instructions by default (without having to click links)
2. Remove "Switch" links, as there's nothing to switch to/from

---

Back to the VNC-available state:

Can you make it where there are only two things to switch to (instead of three)? Basically in-browser (default) with only one switch link to desktop with instructions for both SPICE and VNC listed?

As far as handling special keyboard shortcuts, such as `Ctrl`+`Alt`+`Del`, I think those should be buttons on the top-left, above the VNC viewer.

---

I'll make some wireframes and post them here.

mareklibra commented:
I added https://github.com/cockpit-project/cockpit/pull/5932/commits/c5ba434a8af8c1c7ebfb3a75eda3e74fc9caecb3 fixing the issue with noVnc's unpatchable file.

All noVnc distributables (files) are explicitly listed via symlinks.
The unpatchable keysymdef.js is processed within pkg/Makefile.am .

garrett commented:
## In-Web View (default)

![console-browser-wireframe](https://cloud.githubusercontent.com/assets/10246/24617977/d43c141c-1895-11e7-95c5-1a26d1681a43.png)

Left are the special keyboard shortcuts (`Ctrl`+`Alt`+`Del` and such), center are options (currently unused), and right is the link to switch modes.

Full-screen mode would probably go in the center, where options is currently located. If there are options, then the "View Full Screen" button would go to the left of it (still in the center grouping of widgets).

## Desktop Client View

![console-desktop-wireframe](https://cloud.githubusercontent.com/assets/10246/24617986/d95fafbc-1895-11e7-8e4c-986c8277db88.png)

As the note in the PNG says, the text here on this wireframe is placeholder text.

---

@mareklibra: How well does the above work for you?

mareklibra commented:
@garrett , good work, this looks nice!

Since a VM might have both VNC and SPICE configurred, how can the user decide which one to open?
Please note, there might be even 3rd type in case of windows machines - the RDP, which is planed for a follow-up, AFAIK there's no in-browser variant.

The spice-html5 will be implemented within a follow-up pull request.
How shall the layout look like in the meantime? (so just 3 states are possible: in-browser VNC or desktop SPICE/VNC).

jelkosz commented:
There actually is an in-browser variant for the RDP - but requires IE which by default comes with an activex plugin supporting RDP. In oVirt it is supported, not sure if you want to support something like this in cockpit though...

garrett commented:
@mareklibra: Thanks!

## In-browser viewers

If both VNC and SPICE are enabled on a VM — why would someone choose one over the other in the in-browser UI? Would it make much of a difference which is used in-browser, if they're both configured? Why not just have a default of SPICE and a fallback to VNC (if SPICE is not enabled)? Do they work differently in-browser?

I understand that it's different for desktop viewers and SPICE does offer more than VNC. The above is considering only the in-browser UI.

## Desktop viewers

As for the desktop viewers, I'd say each instruction block should be succinct and should take up 50% of the available width, and wrap around if there are more than 2. (Odds are only 1 or 2 would be enabled at a time, right?) Splitting the width to 33% would make the instructions too narrow.

I don't think I explicitly stated that only the enabled protocols should show up in the desktop viewer instructions, but implied it (with the instructions to connect being tailored per-host).

So I'll say it clearly now: The desktop instructions should only show the available protocols.

## Empty state

I have not made a mockup for this state, but if no protocols are available to connect to the remote machine, then there should be some text stating this, and that SPICE, VNC, and/or RDP (if applicable) should be enabled. 

It should be useful, probably with some way to either enable the protocol(s) directly... or at least a link that would direct them to the correct place within Cockpit to do so (assuming the back button would work in this case at least).

mareklibra commented:
### In-Browser
Simply put, none of in-browser consoles is working perfectly and the user might want to switch them when facing an issue in his particular case.
I'm not comparing protocols, just the in-browser implementations.

Based on our experience (oVirt), the no-vnc works better and is maintained - please see spice-html5 commit activity: https://github.com/SPICE/spice-html5/commits/master
Fot that reason, I would not take it as a default for now.

Within this PR, the no-vnc is implemented and leaving the spice-html5 (or similar better, if any found) for a follow-up.
The code will be prepared for future addition of an in-browser spice, same as for decision mechanism about the default (as you suggested).
If the manual switch for in-browser SPICE/VNC is needed, it seems there's still space left for it.

So I believe, this is not an issue.

### Desktop viewer
There's only single step for basic setup:

    install the remote-viewer

**Optional** step (for better experience with SPICE only):
- install SPICE drivers (with more details)
- for browser on windows only (not applicable for Linux users): install usbdk

@garrett, How does this affect the layout, please?

### Empty state
Right, the user should have an option to define/edit a console for the VM.
But let's postpone implementation for a follow-up and just render a reasonable message for now.

### Console N/A state
A console can be accessed just for a running VM.
How to react for any other VM state?
Recently there's message: 

    Please start the virtual machine to show the graphics console.

mykaul commented:
It doesn't make sense to me to have a 'spice' button (if anything, should be 'Spice') next to 'Restart', for 2 reasons:
1. They are very different one from another.
2. An accidental click on 'Restart' instead of 'Spice' - is there a confirmation, or can you by accident restart your VM?

mareklibra commented:
@mykaul , right, the button is already removed, please have a look at later screenshots in this PR (note, it's still not final design, wip).

Good point with the confirmation. I'll add it within different PR.

mareklibra commented:
Rebased.
Updated according to discussion above taking into consideration
- actual scope of SPICE/VNC related setup
- not-yet implemented spice-html5 support

mareklibra commented:
- general default: in-browser VNC 
![01_vncdefault](https://cloud.githubusercontent.com/assets/17194943/25486950/0857dd00-2b63-11e7-9e0f-62b05a4704f5.png)

- VM with VNC only, user selected desktop viewer
![02_vncdesktop](https://cloud.githubusercontent.com/assets/17194943/25486954/08dbe88e-2b63-11e7-8963-db745f697ed8.png)

-  VM with SPICE only
![03_spiceonly](https://cloud.githubusercontent.com/assets/17194943/25486951/085c84d6-2b63-11e7-88b5-f4aa4121f530.png)

- VM with both VNC and SPICE, in-browser view (means default and no-vnc)
![04_bothspicevnc_default](https://cloud.githubusercontent.com/assets/17194943/25486953/088817cc-2b63-11e7-8787-0509d66cf86f.png)

- VM with both VNC and SPICE, user selected any variant for desktop viewer
![05_bothspicevnc_desktop](https://cloud.githubusercontent.com/assets/17194943/25486952/08875972-2b63-11e7-84ab-7192c325cb60.png)

mareklibra commented:
Tests are failing on:

    KeyError: u'no-vnc has license MPL 2.0 which is not in ./tools/debian/copyright.template'

martinpitt commented:
Right, the MPL-2.0 license blurb needs to be added to tools/debian/copyright.template, as so far we don't have any other MPL-2.0 licensed bower component.

mareklibra commented:
@garrett , any comments, please?

garrett commented:
@mareklibra: It's looking great so far!

## Minor changes

I have a few minor tweaks to suggest (which sadly will have to wait until Monday, as it's the end of the week for me right now and it would take a little too long today to illustrate what should be changed), related to alignment and changing the text a little. This part is mainly cosmetic stuff.

## Functionality changes (connecting via desktop clients)

As for the functionality, how would someone connect to the machines using desktop clients? They cannot just install an app and run it to be connected to the machines; they have to inform the app where to connect. 

What does that look like? (I guess a hostname or IP address and possibly a port and/or a username?) 

This is something I would still like to see on the desktop instructions.

mareklibra commented:
@garrett , regarding the connection details for desktop clients:

The remote-viewer app registers itself to handle (open) files of "application/x-virt-viewer" mime-type.
The cockpit-machines is responsible to generate this file with all the connection details.

From the user perspective, a file download is started when opening the console.
Depending on current browser configuration, the user needs (usually) just to confirm the file to be opened by the suggested application (means the remote-viewer).

martinpitt commented:
I pushed a commit to add MPL-2.0 to the copyright template, sent https://github.com/novnc/noVNC/pull/819 to fix noVNC's license field, and pushed another commit to work around the former.

martinpitt commented:
Not sure why my recent comment https://github.com/cockpit-project/cockpit/pull/5932#discussion_r115192385 is being shown as outdated, but it should be addressed.

martinpitt commented:
Is it possible to split this into two PRs? The first to add "download-style" .vv files for an external viewer, the second for adding the noVNC dependency and an inline graphics console view. This would be much easier to review and land, instead of getting stalled for a long time due to the sheer size and build problems.

martinpitt commented on pkg/machines/components/consoleClientResources.jsx:
```diff
@@ -0,0 +1,58 @@
+    return (
+        <p>
+            {_("For better experience with SPICE, install")}&nbsp;
+            <a href='http://www.spice-space.org/download.html' target='_blank'>{_("SPICE drivers.")}</a>
```
Translatable strings must not be split that way, as word order is often different in other languages. Use something like
```js
cockpit.format(_("For better experience with SPICE, install $0 drivers."),
               "<a href='http://www.spice-space.org/download.html' target='_blank'>SPICE</a>");
```

mareklibra replied:
done

martinpitt commented on pkg/machines/components/consoleClientResources.jsx:
```diff
@@ -0,0 +1,58 @@
+            <p>
+                {_("If console's window did not pop up, please make sure you have the remote-viewer application installed. To do so, either")}
+                <ul>
+                    <li>{_("use package manager:")} <i>yum/dnf install remote-viewer</i></li>
```
The package name is "virt-viewer" in Centos, Fedora and Debian/Ubuntu, and this should add an "apt-get"

mareklibra replied:
done

martinpitt commented on pkg/machines/components/consoleClientResources.jsx:
```diff
@@ -0,0 +1,58 @@
+                    <li>{_("use package manager:")} <i>yum/dnf install remote-viewer</i></li>
+                    <li>{_("or download it from")}&nbsp;
+                        <a href='http://virt-manager.org/download/' target='_blank'>http://virt-manager.org/download/</a>
+                        &nbsp;{_("for various platforms")}
```
Same issue, don't split translatable sentences into multiple parts.

mareklibra replied:
done

martinpitt commented on pkg/machines/include/Orbitron700.ttf:
```diff
@@ -0,0 +1 @@
+../../../bower_components/no-vnc/include/Orbitron700.ttf
```
Shipping 20 static symlinks to various files in noVNC seems really ugly and hard to maintain - tests will very likely not detect if a newer upstream version ships a new or renamed static file which would need to be updated here. Please rather iterate over bower_components/no-vnc/include/ in the Makefile and create the symlinks dynamically.

mareklibra replied:
Done, please have a look at pkg/Makefile.am and webpack.config.js

martinpitt commented on pkg/machines/libvirt.es6:
```diff
@@ -344,7 +381,7 @@ function parseDumpxmlForDisks(devicesElem) {
         }
     }
-
+    
```
stray space

mareklibra replied:
done

martinpitt commented on pkg/machines/manifest.json.in:
```diff
@@ -9,5 +9,6 @@
      }
-  }
+  },
+  "content-security-policy": "default-src 'self' 'unsafe-inline' 'unsafe-eval' data:"
```
This would be the first upstream cockpit module with `unsafe-*`. For what part  is that needed? This should be justified in a comment. Can this be avoided somehow? Is that only necessary for the inline noVNC view, or already for the .vv downloads?

martinpitt commented on webpack.config.js:
```diff
@@ -174,6 +174,28 @@ var info = {
+        "machines/include/Orbitron700.woff",
+        "machines/include/base.css",
+        "machines/include/black.css",
+        "machines/include/blue.css",
```
This repeats that long and changing list another time. Like above, would be better to let a function do it by iterating over include/.

mareklibra replied:
done, see above

garrett commented:
@mareklibra: I'm skeptical about browser helper applications, for several reasons:

1. Different browsers handle helper files for helper apps inconsistently.
2. Operating systems and even desktop environments use differing mechanisms for registering mimetypes. (People access Cockpit from Windows, OS X / macOS, Linux, and so on.)
3. Applications themselves are supposed to be registered to a certain mimetype, but many do not do so.
4. Many different applications can handle VNC/RDP/etc., and someone might want to use a different application from what is registered on their system.
5. Not all helper applications understand helper files.
6. Adding an application when mimetypes fail (for whatever reason) can be a UX nightmare.
7. Depending on system configuration, the helper file might be downloaded to the downloads folder and clutter up someone's system with useless files. (Yes, they're often downloaded to a temporary directory, such as /tmp/, but not always.)
8. Even if all the above points are set up properly and successful, mimetype files still do not consistently work.

Therefore, I strongly believe manual instructions are still needed for connecting to a server — even if a helper file or special URL (such as `vnc://`) might be included. This information includes username, address (hostname or IP), and port.

mareklibra commented:
@garrett , thanks, this more generic approach makes sense.

So I will keep the "help file with mimetype" as it is plus there will be listed key connection details so the user can enter them in any other viewer application.

mareklibra commented:
PR updated:
- rebased
- comments above addressed
- "view with desktop client" is improved

Both parts (w/o no-vnc) are still together. I can split if needed, but the in-browser part was requested to land this feature previously. In addition, I did some "bugfixing" in the later no-vnc part which are related to the first commit. Anyway, commits are squashed and reordered - I can split if needed but it's extra work to move some of the later changes to the first commit.

Recent screenshots:
![01_desktopconsole](https://cloud.githubusercontent.com/assets/17194943/25901871/4ea4be42-3598-11e7-9688-814e2aab07c9.png)

![02_desktopconsolesingle](https://cloud.githubusercontent.com/assets/17194943/25901872/4eba7804-3598-11e7-88b7-46ac370e7dfa.png)

cockpituous commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Virtual Machines: Display graphics console](https://trello.com/c/EEnqhlUz/503-virtual-machines-display-graphics-console)

cockpituous commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Virtual Machines: Display graphics console](https://trello.com/c/EEnqhlUz/503-virtual-machines-display-graphics-console)

mareklibra commented on pkg/machines/components/graphicsConsole.jsx:
```diff
@@ -168,9 +168,17 @@ class GraphicsConsole extends React.Component {
 
         this.onBrowserVnc = this.onBrowserVnc.bind(this);
+        this.onDesktopConsole = this.onDesktopConsole.bind(this);
+    }
```
done

garrett commented:
@mareklibra: I'm running the branch here locally — and I know it's a work in progress — but:

1. I'm wondering how I can run SPICE & VNC in the browser (as they don't seem to work here).
2. Should it download a file in download mode by default? (It doesn't. And I think we want to revisit whether or not if should download as well.)
3. I only see SPICE... but not VNC.
4. Clicking the button does nothing.
5. What does "Please make sure the SPICE driver is installed for better experience." mean? I visit the link and I don't know what I should do based on that page.

Note: Using the IP address works fine here. I can connect to my VMs with "Remote Desktop Viewer" (aka: Vinagre). Thankfully that information was provided, so I could use a different app (the other is installed here, but doesn't do anything by default). :smile:

mareklibra commented:
@garrett , recently just the no-vnc as an in-browser console is provided.
The spice-html5 is left for a potential follow-up.

As requested in some previous comment, the in-browser console is the default.
In other words, recently "the default" means: if a VM has VNC, then in-browser no-vnc, otherwise SPICE desktop.

1.  The VNC in-browser console shall be working. If not, it might be bug.

2., 4.  What browser do you use? I found various issues with CSP (content security policy), behavior among browsers (FF/Chrome) varies for the same configuration. I guess you are facing it.
Please check browser console for an error message (not necessarily from the time of console retrieval).
I will further try to improve that or at least provide a workaround I found (which is not nice but seems to be working).

Anyway, the expected behavior is to automatically start a .vv file download if remote-viewer shall be invoked.
That's the reason for the rendered "Open the just downloaded .vv file ..." message.

3. Is your VM defined with both types? To check: 
 - virsh -r dumpxml [VM_NAME]
 - look for the <graphics> elements, both type='spice' and 'vnc' should be present

5. No problem with that. The "Reconnect" is used since the .vv file download should be started automatically (see 2.4. above)

6. Right, this should be rephrased, content of SPICE download page is not clear at the first glance. 

It's related to the SPICE console only. The SPICE comes with **optional** OS drivers to improve either performance or extend features (they vary among platforms). The referenced SPICE download page should provide sufficient guiding.

Namely, Windows users might be interested in the QXL, UsbDk or guest-tools.
Linux users might be looking for the GTK+ Widget or QXL drivers.

Their installation is out of Cockpit's scope. I just wanted to provide external reference.

Any idea how to make it more clear from Cockpit perspective? I would not avoid it completely, since e.g. the Windows user might be lost is usb redirection is not working out of the box.

mareklibra commented:
@garrett , I just pushed aeb8783 , it might fix your issue with skipped automatic .vv file download. Please let me know.

garrett commented:
@mareklibra:

I use Firefox as my main browser. I have Chromium, Chrome, and Epiphany also installed for various reasons (testing, Bluejeans, web apps as desktop apps, etc.).

---

I don't know how to check (or enable) VNC in addition to SPICE — you're assuming I know what the VMs are called according to `virsh`, which doesn't make it easy to get the names (after looking at help for a few minutes, it's _not_ obvious). :wink:

(I'm assuming VNC isn't enabled, as it doesn't show up in the console tab when I'm using your branch.)

---

My point with the SPICE link is simply:

_I followed the link and I had no clue what I should do at that page._

I don't know if I have optimal stuff installed or not, how to check, or even how to install or set it up. The SPICE page that is linked, while huge, is not useful at all as it doesn't seem to cover any of that. (If it does, it's lost in the huge amount of text.)

---

As far as the reconnect/connect and auto-downloading files issues — I was implying in a few places and even outright said (at the end of item # 2) we should not auto-download a file. The only times I've seen auto-downloading used were:

1. **Download redirect pages** (like on SourceForce) — but these only show up when someone has explicitly asked for a file.
2. **Malware** (usually the form of an .exe)
3. **Java webstart apps**, years ago.

...and that's it. Perhaps others here have had other experiences? Most of these aren't really great experiences (waiting for a download when you've already asked for one, getting malware, or, well, using Java apps :wink:).

One aspect I did not previously mention: Firefox disrupts flow as well, as it pops up a model dialog for downloaded files that are to be used with helper applications. Even still… for Firefox, Chrome, or any other browser: clicking view in desktop link or simply switching to the VM console tab within Cockpit shouldn't spawn an external application. It should be something that's specifically asked for — it should be specifically requested with one of the connect buttons.

**Short summary**: Downloading files should only happen from a UI where it's obvious a file will be downloaded (even if it's a launcher file).

mareklibra commented:
@garrett , "edit" functionality will be added later but definitely planed. The first step towards that is the create VM feature, which I sent for a discussion to the mailing list.

As you mentioned, the "Consoles" recently just reflects the configuration of existing VM.

If you want to get more details via `virsh`, the VMs are referenced via their "domain name". It is same as we use as for "VM name" within cockpit-machines.

The "Desktop view" will be changed to require extra step (click) to download the .vv file.

garrett commented:
@mareklibra: FWIW:

There are several remote viewers that Fedora, RHEL, CentOS, Debian, and Ubuntu ship.

In Fedora alone, I see:

1. **virt-viewer**, called "**Remote Viewer**"
2. **vinagre**, called "**Remote Desktop**"
3. **gnome-boxes**, called "**Boxes**" (this lets you run local VMs as well as connect to remotes over SPICE, VNC, RDP)

(The name each is called is what they show up as in the .desktop files for their launcher name. It's also what's seen in GNOME-Software.)

All of these support SPICE and VNC; some also support RDP (Windows remote desktop). The latter two let you save machines so you basically have 1 or 2 clicks to get back to them. This is one of the many reasons why it's a good idea to expose the IP address. 😄

I'm guessing that the .vv files are only supported in `virt-viewer`, sadly. It would be nice to have the app of choice handle the connection.

Also, there are probably more (especially in Debian, I would guess); these are just the ones exposed in GNOME-Software in Fedora.

garrett commented:
@mareklibra: I've made a mockup based on what we talked about and what I was trying to explain in the video chat. :wink:
[![console desktop mockup](https://raw.githubusercontent.com/cockpit-project/cockpit-design/master/virtual-machines/console-desktop.png)](https://raw.githubusercontent.com/cockpit-project/cockpit-design/master/virtual-machines/console-desktop.png)

mareklibra commented:
@garrett , thanks, that looks great and meets the requirements! :-)

The address might vary between protocols, so I will deduplicate if possible or provide specific otherwise.
Later on, if there is some real use case to explicitly select the protocol for the "Launch Remote Viewer", it can be added along other potential options.

mareklibra commented:
Rebased and updated according to last design.

Ports are displayed accroding to actual VM configuration, so some subset of:
- SPICE, TLS SPICE
- VNC, TLS VNC

If address is same for both SPICE and VNC, then it's deduplicated. Or protocol-specific is rendered otherwise.

User's selection of "Switch to ..." is stored to browser's local storage so user's preference is taken into account when opening any-VM's Console subtab next-time.

Screenshots:
![01_defaultinbrowser](https://cloud.githubusercontent.com/assets/17194943/26247276/7486a27c-3c9d-11e7-84b9-fe74ef23c668.png)

![02_desktop](https://cloud.githubusercontent.com/assets/17194943/26247279/783ebcb0-3c9d-11e7-8260-d15ec3ac6725.png)

![03_desktopmoreinfo](https://cloud.githubusercontent.com/assets/17194943/26247281/7bcdd596-3c9d-11e7-9ea5-303640657d2a.png)

mykaul commented:
@garrett  - I wonder if the manual connection part can be a string that remote-viewer and friends will be able to use. 
So you just have to copy-paste a single line to be used in them.
As is, right now, you have to either enter manually or copy-paste multiple times.

mareklibra commented:
Rebased only

martinpitt commented on pkg/Makefile.am:
```diff
@@ -13,6 +13,19 @@ playground_DATA = \
+
+machines_DATA = \
+	dist/machines/include/keysymdef.js \
+	machines_novnc_symlinks \
```
automake rules are not that magical - i. e. accepting a phony rule name here, instead of plain file names:
```
$ make install DESTDIR=`pwd`/i
 /usr/bin/install -c -m 644 dist/machines/include/keysymdef.js ./machines_novnc_symlinks '/home/martin/upstream/cockpit/i/usr/share/cockpit/machines'
/usr/bin/install: cannot stat './machines_novnc_symlinks': No such file or directory
```

This needs to be spelled out like this rather: `machines_DATA = [...] $(shell find ...)`.

martinpitt commented on tools/webpack-make:
```diff
@@ -131,6 +131,9 @@ function generateDeps(makefile, stats) {
             !endsWith(output, ".map") &&
+            !endsWith(output, ".ttf") &&
+            !endsWith(output, ".woff") &&
+            !endsWith(output, ".map") &&
```
`.map` is already present three lines above, so this is redundant.

martinpitt commented:
I tested this on Fedora 25's libvirt/virt-manager. I launched a VM (running Debian sid with LXDE) and see it in Cockpit Machines. The external viewer (virt-viewer) works fine (.vv dowload etc.). With libvirt's default configuration there is no VNC available, only spice, so the inline viewer isn't offered. After adding a "Display VNC" device (which is not at all easy to find), the inline viewer is broken:

![screenshot](http://picpaste.com/pics/vnc-LXxccKH7.1496055929.png)

The log:
```
Firefox can’t establish a connection to the server at wss://localhost:9090/cockpit/channel/7630c523071c2a053b7a8f6823f423c28ac2cf3b18fe15359cf10dbe6320c83e?eyJwYXlsb2FkIjoic3RyZWFtIiwicHJvdG9jb2wiOiJiaW5hcnkiLCJhZGRyZXNzIjoiMTI3LjAuMC4xIiwicG9ydCI6NTkwMSwiYmluYXJ5IjoicmF3In0=.  websock.js:292:30
Attempt to set a forbidden header was denied: Connection  564651399-bc_manager_core.js:112:107
Firefox can’t establish a connection to the server at wss://localhost:9090/cockpit/channel/7630c523071c2a053b7a8f6823f423c28ac2cf3b18fe15359cf10dbe6320c83e?eyJwYXlsb2FkIjoic3RyZWFtIiwicHJvdG9jb2wiOiJiaW5hcnkiLCJhZGRyZXNzIjoiMTI3LjAuMC4xIiwicG9ydCI6NTkwMSwiYmluYXJ5IjoicmF3In0=.  websock.js:292:30
[Exception... "Component returned failure code: 0x80070057 (NS_ERROR_ILLEGAL_VALUE) [nsIAnnotationService.setPageAnnotation]"  nsresult: "0x80070057 (NS_ERROR_ILLEGAL_VALUE)"  location: "JS frame :: resource://app/modules/DownloadsCommon.jsm :: onDownloadChanged :: line 801"  data: no]  (unknown)
Attempt to set a forbidden header was denied: Connection  564651399-bc_manager_core.js:112:107
Object { channel: channel(), event_target: event_target(), extend: extend(), manifests: Object, utf8_encoder: utf8_encoder(), utf8_decoder: utf8_decoder(), base64_encode: base64_encode(), base64_decode: base64_decode(), kill: kill(), hint: hint(), 44 more… } Object { wait: ensure_transport(), inject: inject(), filter: filter(), close: close(), origin: "http://localhost:9090", options: Object, uri: calculate_url(), control: control(), application: application(), csrf_token: "7630c523071c2a053b7a8f6823f423c28ac…", 1 more… }  vnc.html:224:13
channel 5901  vnc.html:225:13
 Msg: noVNC ready: native WebSockets, canvas rendering  util.js:218:43
  util.js:218:43
Firefox can’t establish a connection to the server at wss://localhost:9090/cockpit/channel/7630c523071c2a053b7a8f6823f423c28ac2cf3b18fe15359cf10dbe6320c83e?eyJwYXlsb2FkIjoic3RyZWFtIiwicHJvdG9jb2wiOiJiaW5hcnkiLCJhZGRyZXNzIjoiMTI3LjAuMC4xIiwicG9ydCI6NTkwMSwiYmluYXJ5IjoicmF3In0=.  websock.js:292:30
WebSocket on-error event  util.js:218:43
WebSocket on-close event  util.js:218:43
 Msg: Server disconnected (code: 1006)  util.js:220:43
	Util.init_logging/Util.Error http://localhost:9090/cockpit/@localhost/machines/include/util.js:220:43
	_updateState http://localhost:9090/cockpit/@localhost/machines/include/rfb.js:456:17
	_fail http://localhost:9090/cockpit/@localhost/machines/include/rfb.js:526:13
	RFB/< http://localhost:9090/cockpit/@localhost/machines/include/rfb.js:209:17
	open/this._websocket.onclose< http://localhost:9090/cockpit/@localhost/machines/include/websock.js:320:17
  util.js:218:43
Resizing noVNC  vnc.html:101:17
```

This happens in Firefox and Chrome. This might be a side effect of running from a built tree and a ~/.local/share/cockpit → ~/upstream/cockpit/dist symlink (as package building is currently broken)? How are you testing this?

martinpitt commented on pkg/machines/components/desktopConsole.jsx:
```diff
@@ -0,0 +1,219 @@
+                <MoreInformationInstallVariant os='RHEL, CentOS' command='sudo yum install virt-viewer'/>
+                <MoreInformationInstallVariant os='Fedora' command='sudo dnf install virt-viewer'/>
+                <MoreInformationInstallVariant os='Ubuntu, Debian' command='sudo apt-get install virt-viewer'/>
+                <MoreInformationInstallVariant os='Windows' innerHtml={downloadMsg}/>
```
Not a blocker, just a suggestion: All of this is essentially a constant HTML `<ul>` - why the complexity of a separate component for this? Seems simpler to understand and maintain to just spell it out here?

mareklibra replied:
I tried to avoid copy-paste, but I can "expand".

mareklibra replied:
@martinpitt , please let me know if i shall expand. I think recent implementation is better maintainable, but I can expand anyway ...

martinpitt replied:
It's not a big deal, if you prefer it that way then let's just keep it as it is.

Thanks for the other fixes! I'll re-review tomorrow morning.

Is there something that can be done about http:// not working with the inline viewer? Such as not offering it in the first place if SSL is off, instead of showing a broken widget?

martinpitt commented on pkg/machines/components/desktopConsole.jsx:
```diff
@@ -0,0 +1,219 @@
+
+const _ = cockpit.gettext;
+
+const MoreInformationInstallVariant = ({ os, command, innerHtml }) => {
```
As I'm learning react still and am interested: Is this "declare a const lambda expression" style in any way better or more performant than the standard and [documented](https://facebook.github.io/react/tutorial/tutorial.html#functional-components) functional component syntax `function MoreInformationInstallVariant(os, command, innerHtml)`?

mareklibra replied:
It's just an ES6 syntax for the same thing. Differences are small (see https://developer.mozilla.org/cs/docs/Web/JavaScript/Reference/Functions/Arrow_functions). I just got used to it.

Regarding passing params as object - it's handy and extensible. And sometimes less error-prone.

martinpitt commented on pkg/machines/components/desktopConsole.jsx:
```diff
@@ -0,0 +1,219 @@
+    if (!display) {
+        display = vm.displays.vnc;
+    }
+    const onLaunch = () => onDesktopConsole(display)
```
Looks odd to not have a terminating semicolon here, although it seems to work anyway. (Same issue in further const declarations below)

mareklibra replied:
Good catch, accidentally missed. It's another ES6 syntax. I will add the semicolon.

martinpitt commented on pkg/machines/components/graphicsConsole.jsx:
```diff
@@ -0,0 +1,198 @@
+const VmNotRunning = () => {
+  return (
+      <div>
+          {_("Please start the virtual machine to access its graphics console.")}
```
This does not seem to work properly -- if I define a VM and not start it, or shut it down, I see "No graphics console is defined.." from above, not this one. Also, this seems to be a bit excessively componentized - just put the string into `render()` where you need it?

mareklibra replied:
Not a recent use case, but it would be nice to have a "solution" here. 
Anyway I will have a look at the flow and prefer this message over no-console state

mareklibra replied:
done

martinpitt commented on pkg/machines/components/graphicsConsole.jsx:
```diff
@@ -0,0 +1,198 @@
+const setDefaultBrowserMode = (inBrowser) => {
+    const value = inBrowser ? 'true' : 'false';
+    logDebug('Default Browser Console mode set to: ', value);
+    window.localStorage.setItem('MACHINES_CONSOLE_IN_BROWSER_DEFAULT', value);
```
Would be nicer to keep this as state in `GraphicsConsole` (or perhaps `SwitchConsole`)?

mareklibra replied:
There was requirement to make this choice persistent.
For completeness, there will be more options related to consoles persisted this way (means in the browser's storage)

martinpitt commented on pkg/machines/manifest.json.in:
```diff
@@ -9,5 +9,6 @@
      }
-  }
+  },
+  "content-security-policy": "default-src 'self' 'unsafe-inline' data:"
```
What's left that requires unsafe-inline? No other cockpit-shipped page currently needs this, so this requires some justification (at least a comment). Can this be avoided somehow?

mareklibra replied:
Please suggest what's the best place to document this.
So far I put following to the PR description (see above):

    The cockpit-machines newly sets
    "content-security-policy": "default-src 'self' 'unsafe-inline' data:"
    to make the no-vnc and download of dynamically generated files (the .vv) working.

martinpitt replied:
Ah, having it in the commit description (not just the PR) seems okay. Unfortunately JSON files can't have comments.

martinpitt commented on pkg/machines/vnc.html:
I believe inline <style> violates CSP? Can this move into the .less file? There are also a bunch of `style=` properties below which would have the same problem

mareklibra commented on pkg/machines/vnc.html:
This vnc.html comes from no-vnc example which I expect is/will be maintained by no-vnc developers.
To keep changes to this file at minimum might help maintenance in the future.

martinpitt commented:
Missed you on IRC, you asked about uglifying keysymdef.js. I think it would be best to include the generated file into the tarball ("dist" it) so that uglifyjs is not necessary during building packages.

martinpitt commented on pkg/machines/vnc.html:
But it should be fine to put that into a .css file and just <link> it?

mareklibra commented on pkg/machines/vnc.html:
Done - moved to new vnc.css file

mareklibra commented:
RPM build shall be fixed now, dist tgz contains already preprocessed no-vnc's keysymdef.js, so the "uglify" is not called at rpm build.

martinpitt commented:
Also, this needs integration tests before it can actually land. Presumably the no-VNC widget won't work in phantomjs, but they should at least cover the external methods: general layout, .vv download (and verifying its contents) and that the "manual connection" data is correct and displayed. Do you want to look into this, or want/need some help?

martinpitt commented:
I started a TODO list in the description, please feel free to amend/update.

martinpitt commented:
The recent fixes LGTM, tests now (by and large) run, and it now correctly says "please start the machine" in the console tab if it's not running. For some reason keyboard input still doesn't work in the inline viewer -- it did yesterday morning, but didn't when I tried again (with the same code, just another VM boot), and still doesn't workw with your updates. Does that work reliably for you?

The remaining issues are in a checklist in the description. Thanks, this is making nice progress!

mareklibra commented:
I've white-listed dist/machines/include/logo.js in tools/check-dist since this is 3rd party source code

mareklibra commented:
@martinpitt , regarding the keyboard input: seems like the no-vnc canvas had no focus. I changed to focus the canvas after its load.

mareklibra commented:
wss/ws for no-vnc chanel is set according to https/http protocol being used

mareklibra commented:
Integration test added (check-machines).

Does anyone have suggestion how to test the .vv file download and content with phantomjs?

martinpitt commented:
>  test the .vv file download and content with phantomjs

I'm not a phantomjs pro yet either, but I don't think it's as good as providing file dialogs. Is it possible to just assert that the button/link exists with a proper target URL, then `m.execute("curl <url>")` that and verify its contents that way?

mareklibra commented:
Integration test is finished - content of the .vv is checked right before actual download.
It's a complex question whether the download itself works, especially due to CSP. So this test assures validity of content.

martinpitt commented:
@mareklibra : I tested the current version again, and it has lost the buttons for Run/Shutdown/Restart. They are neither on the Overview nor the Console tab any more.

![screenshot](http://picpaste.com/pics/vnc-fCkfl4DF.1496399546.png)

mareklibra commented:
@martinpitt , thanks for noticing that. Maybe incorrect rebase ... I'll have a look

martinpitt commented:
@mareklibra : My woes seem to be something else -- now it doesn't work with master or 141 any more either, the VMs are always shown as stopped even when running and there are no run/shutdown buttons. virt-manager and virsh work fine, though.

martinpitt commented:
My woes were due to me changing the display language to German some days ago to test something else. Cockpit's VM UI is broken then; it might be issue #5967 (I asked for confirmation), if not I'll file a new issue. So I'll continue testing.

martinpitt commented:
The failed [check-machines TestMachines.testBasic failure on ubuntu-1604](https://fedorapeople.org/groups/cockpit/logs/pull-5932-20170607-110505-43106283-verify-ubuntu-stable/log.html#93) seems happen reliably:
```
Unexpected journal message 'audit: type=1400 audit(1496834273.336:14): apparmor="DENIED" operation="open" profile="libvirt-80460ea5-12d9-4ecc-9507-b00a935a0bc4" name="/etc/gss/mech.d/" pid=5548 comm="qemu-system-x86" requested_mask="r" denied_mask="r" fsuid=64055 ouid=0'
```
I'll see to reproduce this locally and add a quirk for it.

martinpitt commented:
I filed the AppArmor issue downstream (https://launchpad.net/bugs/1696471) and created a known issue in #6901. I pushed a new commit to add the quirk, which should now make the tests happy. I also rebased on top of current master for easier local testing, so I needed to force-push. @mareklibra , you need to hard-reset your tree  for this, sorry.

martinpitt commented:
Curiously, there's an SELinux  denial in the same test on Fedora 26, but on `/proc`/:
```
Unexpected journal message 'audit: type=1400 audit(1496866431.283:268): avc:  denied  { search } for  pid=1474 comm="qemu-system-x86" name="847" dev="proc" ino=20523 scontext=system_u:system_r:svirt_tcg_t:s0:c86,c242 tcontext=system_u:system_r:virtd_t:s0-s0:c0.c1023 tclass=dir permissive=0'
```
Not sure if that's a flake or due to this PR, I'll retry and check.

martinpitt commented:
Seems the SELinux `/proc` denial was a flake, didn't happen in two further retries. However, `testConsoles (check_machines.TestMachines)` fails awfully often with timing out for waiting for the console, and the VM is in state "paused":

![screenshot](https://fedorapeople.org/groups/cockpit/logs/pull-5932-20170608-060606-7fbe87b0-verify-fedora-26/TestMachines-testConsoles-FAIL.png)

This is most likely unrelated to this PR as well,  but @mareklibra , does that ring a bell? I don't see anything standing out in the [journal](https://fedorapeople.org/groups/cockpit/logs/pull-5932-20170608-060606-7fbe87b0-verify-fedora-26/TestMachines-testConsoles-10.111.116.48-FAIL.log)

martinpitt commented:
@mareklibra: PR #6814 moves away from bower and to only using npm. https://www.npmjs.com/package/novnc-node is slightly older (0.5.3) than https://libraries.io/bower/no-vnc (0.6.2) - is that sufficient, or do we need to get the "upstream" novnc registered? Or can/should we just directly specify the git branch at https://github.com/novnc/noVNC and use the latest tag?

mareklibra commented:
@martinpitt , regarding the SELinux: I found that starting nested VMs on Cockpit's test machines is unreliable. The SELinux is not the only reason why virt-install undeterministically fails.

For this reason, there are already "if-branches" in check-machines dealing with unstarted (means usually paused) nested-test-VM. This environment-related issue shall not happen, but the test deals with this falsely-fail gracefully by logging and skipping it at least.

I will add similar handling to the new _testConsoles_.

martinpitt commented:
@mareklibra: I don't like skipping much I must say, I'd rather retry the test; it could otherwise happen that due to some real bug the test always gets skipped and we never notice. Could it retry instead?

mareklibra commented:
@martinpitt , noVNC versions: when looking at https://github.com/novnc/noVNC/releases , the 0.5.3 is already quite old and there are even security fixes on the way towards 0.6.2.

The npm package is maintained by https://github.com/eface2face/noVNC-node . From my brief observation, it does not look like reliable solution to me.

Releases from https://github.com/novnc/noVNC/releases seem to contain the important `include` directory, effectively the only thing we need.
Is there a clean way how to download it during autogen.sh ? And how to deal with licence, etc.

mareklibra commented:
@martinpitt , restart of falsely-failed tests: once the nested-VM creation fails (not actual part of the test, just its preparation), re-invocation of virt-install will not help much. The only way might be to start over - this must be handled by code outside the _testConsoles_ or _testBasic_. Is there some config option for that?

In other words, it would retriger the test as recently done manually when a PR looks more-or-less good to merge but just waiting for all-green.

Automatic retriggering (or sort of it) might have negative consequences on capacity of the test environment.

So hard to choose :-)

martinpitt commented:
@mareklibra : Agreed, so let's not use that facebook fork. But I don't see any need for autogen.sh hacks, with
```diff
--- a/package.json
+++ b/package.json
@@ -29,6 +29,7 @@
     "less": "~2.6.0",
     "less-loader": "~2.2.3",
     "ng-cache-loader": "0.0.16",
+    "noVNC": "https://github.com/novnc/noVNC/archive/v0.6.2.tar.gz",
     "po2json": "~0.4.1",
     "promise": "~7.1.1",
     "raw-loader": "~0.5.1",
```
you get a `node_modules/noVNC/include/` which looks exactly like `bower_components/no-vnc/include/`, so that should be alright? Note that I'm not saying that this should be changed *right now* (I think PR #6814 should land  before switching noVNC from bower to npm), just that we need to prepare for it.

But the `testConsoles` failure on Fedora 26 is a blocker for landing this PR. I agree, let's use the same hack as in `testBasic` for now. Can you please update that? Thanks!

mareklibra commented:
With recent force-push, I've fixed size of the noVNC widget - browser window's vertical scrollbar is used instead. Horizontal scrollbar is still within the widget.

The `testConsoles` in check-machines has newly `if state == "running"` check to workaround the issue.

martinpitt commented:
I see you already added the "paused" check. Thanks for fixing the VNC widget size, this looks nice now! I ticked that todo box. So, once the tests are green this is ready to land from my POV. Great work!

mareklibra commented:
Thanks! :-)

martinpitt commented:
I squashed the commits a bit and pushed. Yay!

martinpitt commented:
@mareklibra: I'll handle the bower→npm migration in #6814; I'll let you know if I need a hand.

garrett commented:
@mareklibra: I've filed 2 small enhancements. Once is minor text adjustment (#6940) and the other is a feature to scroll the VM console into view if it's partially cropped and the scroll region would allow for it (#6939).

Otherwise, everything still looks good to me, based on the blog post @ http://cockpit-project.org/blog/cockpit-142.html — thanks for the screenshots and videos there!
